pkg/skaffold/jib/jib.go

```diff
@@ -17,64 +17,173 @@ limitations under the License.
 package jib
 
 import (
+	"encoding/json"
 	"os"
 	"os/exec"
+	"path/filepath"
 	"sort"
+	"strings"
+	"sync"
+	"time"
 
 	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/util"
 	"github.com/karrick/godirwalk"
 	"github.com/pkg/errors"
 	"github.com/sirupsen/logrus"
 )
 
-func getDependencies(cmd *exec.Cmd) ([]string, error) {
-	stdout, err := util.RunCmdOut(cmd)
-	if err != nil {
-		return nil, err
+// pathMap contains fields parsed from Jib's JSON output
+type pathMap struct {
+	// Build lists paths to build definitions that trigger a call out to Jib to refresh the pathMap, as well as a rebuild, upon changing
+	Build []string `json:"build"`
+
+	// Inputs lists paths to build dependencies that trigger a rebuild upon changing
+	Inputs []string `json:"inputs"`
+
+	// Results lists paths to files that should be ignored when checking for changes to rebuild
+	Results []string `json:"ignore"`
+}
+
+// filesLists contains cached build/input dependencies
+type filesLists struct {
+	// PathMap saves the most recent output of the Jib Skaffold files task/goal
+	PathMap pathMap
+
+	// BuildFileTimes keeps track of the last modification time of each build file
+	BuildFileTimes map[string]time.Time
+}
+
+// watchedFiles maps from project name to watched files
+var watchedFiles = sync.Map{}
+
+// getDependencies returns a list of files to watch for changes to rebuild
+func getDependencies(cmd *exec.Cmd, projectName string) ([]string, error) {
+	var dependencyList []string
+	f, ok := watchedFiles.Load(projectName)
+	if !ok {
+		f = filesLists{}
+	}
+	files := f.(filesLists)
+
+	template := files.PathMap
+	if len(template.Inputs) == 0 && len(template.Build) == 0 {
+		// Make sure build file modification time map is setup
+		if files.BuildFileTimes == nil {
+			files.BuildFileTimes = make(map[string]time.Time)
+		}
+
+		// Refresh dependency list if empty
+		if err := refreshDependencyList(&files, cmd); err != nil {
+			return nil, errors.Wrap(err, "initial Jib dependency refresh failed")
+		}
+
+	} else if err := walkFiles(&template.Build, &template.Results, func(path string, info os.FileInfo) error {
+		// Walk build files to check for changes
+		if val, ok := files.BuildFileTimes[path]; !ok || info.ModTime() != val {
+			return refreshDependencyList(&files, cmd)
+		}
+		return nil
+	}); err != nil {
+		return nil, errors.Wrap(err, "failed to walk Jib build files for changes")
+	}
+
+	// Walk updated files to build dependency list
+	if err := walkFiles(&files.PathMap.Inputs, &files.PathMap.Results, func(path string, info os.FileInfo) error {
```

Comment:
If `pathMap` is embedded into `filesLists`, you can then write `&files.Inputs` instead of `&files.PathMap.Inputs`